### DISCUSSION TITLE
Strange behavior of json?

### DISCUSSION THREAD

**Discussion by vlulla** [duckdb/duckdb#11896] [General]
Consider this example table:

```sql
create table example(j json);
insert into example values
         ('{  "family": "anatidae", "species": ["duck", "goose", "swan", null] }'),
         ('{"family":"anatidae","species":["duck","goose","swan",null]}') -- only whitespaces are different
;
```

And, here are a couple of queries:

```sql
select distinct j.family, j.species from example; -- returns one row;
select distinct j from example; -- surprisingly this return both rows!
```

Isn't this behavior inconsistent? Or, am I mistaken in finding _this_ behavior inconsistent?

**Comment by Tishj** [comment]
I would be interested to see this further nested
Because as it stands the "family" and "species" objects dont have differing whitespace

I imagine when they do that the first query will also yield two rows

**Reply by vlulla** [reply]
I don't think i understand your comment. Can you please give an example of the difference you are expecting if "family" and "species" objects had differing whitespace?

On a separate note, i was surprised to find that the first query (with `distinct` clause) does not work with PostgreSQL (at least v16.2)!

**Reply by Tishj** [reply]
Ah it seems I misread this on my phone, they do have differing white spaces:
`"family": "anatidae"`
`"family":"anatidae"`

But I wonder what happens when family is also a complex object, i.e `"family": {"key": "value"}`

**Comment by Tishj** [comment]
What I think happens is that a very shallow string compare is performed on the object

So it's not checking every sub-object recursively
I'm not sure if this is expected or not
Laurens should have a better idea on that

**Reply by vlulla** [reply]
Hmm...i read. Shouldn't this be a "json" compare instead of "string" compare? I find the result unexpected because I have explicitly defined column `j` as `json` and not `varchar`.

**Comment by lnkuiper** [comment]
Hi @vlulla, thanks for reporting this! This is currently expected behavior. DuckDB does not consider those two equal because it simply performs a string comparison when comparing any two JSONs with each other.

We could potentially prevent this by "minifying" any JSON inserted into DuckDB. I could see us changing this at some point in the future, but it's not a high priority for us.

**Reply by soerenwolfers** [reply]
The documentation might benefit from a warning added to the current "The JSON extension makes use of the JSON logical type. The JSON logical type is interpreted as JSON, i.e., parsed, in JSON functions rather than interpreted as VARCHAR". I opened https://github.com/duckdb/duckdb-web/issues/2808

**Reply by vlulla** [reply]
I agree that a warning would be very helpful. I was surprised by observing the incongruence from the behavior described on the exact line that @soerenwolfers mentioned! Thanks for your consideration.
